src/compiler/utilities.ts

```diff
@@ -2056,19 +2056,51 @@ namespace ts {
             }
             return AssignmentDeclarationKind.ObjectDefinePropertyValue;
         }
-        if (expr.operatorToken.kind !== SyntaxKind.EqualsToken ||
-            !isPropertyAccessExpression(expr.left)) {
+        if (expr.operatorToken.kind !== SyntaxKind.EqualsToken) {
             return AssignmentDeclarationKind.None;
         }
-        const lhs = expr.left;
-        if (isEntityNameExpression(lhs.expression) && lhs.name.escapedText === "prototype" && isObjectLiteralExpression(getInitializerOfBinaryExpression(expr))) {
-                // F.prototype = { ... }
-                return AssignmentDeclarationKind.Prototype;
+        if (isPropertyAccessExpression(expr.left) || isElementAccessExpression(expr.left)) {
+            const lhs = expr.left;
+            if (isEntityNameExpression(lhs.expression) && getElementOrPropertyAccessName(lhs) === "prototype" && isObjectLiteralExpression(getInitializerOfBinaryExpression(expr))) {
+                    // F.prototype = { ... }
+                    return AssignmentDeclarationKind.Prototype;
+            }
+            return getAssignmentDeclarationPropertyAccessKind(lhs);
+        }
+        return AssignmentDeclarationKind.None;
+    }
+
+    // TODO: Signed numeric literals?
+    /* @internal */
+    export function getElementOrPropertyAccessImmediateRHS(node: PropertyAccessExpression | ElementAccessExpression): Identifier | StringLiteralLike | NumericLiteral | ElementAccessExpression | undefined {

```

Comment:
"immediate rhs" is kind of a confusing name when the properties are `name` and `argumentExpression`. I would lean toward `name` instead, although I'm not sure.

Reply:
Oh, it's already taken. Then `nameExpression`?